lua/entities/gmod_wire_expression2/core/custom/cfc_npc.lua

```diff
@@ -0,0 +1,129 @@
+CFCE2Lib.RegisterExtension( "cfc_e2_lib", true )
+
+-- Made by StrawWagen
+
+-- Setting up local functions 
+
+local function ValidPly( ply )
+    if not ply or not ply:IsValid() or not ply:IsPlayer() then
+        return false
+    end
+    return true
+end
+
+local function ValidNpc( npc )
+    if not npc or not npc:IsValid() or not npc:IsNPC() then
+        return false
+    end
+    return true
+end
+
+
+local function hasAccess(ply)
+    if ply:IsAdmin() then
+        return true
+    end
+    return false
+end
+
+
+-- Admin only npc functions
+
+e2function void entity:npcSetMaxHealth( number maxhealth )
+    if not ValidNpc( this ) then return end
+    if not ValidPly( self.player ) then return end
```

Comment:
In alot of these you could do the validPly check in hasAccess
specifically in any function where self.player is only used to check access

Reply:
i prefer keeping the functions seperate, if i were to combine them it would mean either making a new function or making one sorta redundant, i can change it if it's that big of a problem but i think we can probably get away with leaving it as-is

Reply:
It's the same checks every time. Later, if you decide that there needs to be another check added, you'd need to go through every function and replicate that change. This can lead to mistakes, errors, or unexpected behavior.

To keep code DRY is to keep it maintainable.

Further, in this case, you're rejecting functionality if the player is invalid. That sounds like granting conditional access to me.

```lua
local function hasAccess( ply, requiresAdmin )
    if not IsValid ( ply ) then return false end -- IsValid should be sufficient

    if not requiresAdmin then return true end
    if not ply:IsAdmin() then return end
end
```

Reply:
I could see the value of keeping  the validPly check separate if the function used self.player
but in this case `self.player` is only relevant to the hasAccess function